### PR TITLE
planner: check index valid while forUpdateRead

### LINKED ISSUE #21498: RC transactions may see inconsistent data when there are concurrent DDLs
## Bug Report

**The reason**:  In the current implementation of RC transaction, it fetches a latest ts to read data for each read statement. However it always uses the schema at the transaction's startTS (if I'm correct). So if the schema changed while the transaction is running, the transaction may see data that doesn't match the definition of the schema.

### 1. Minimal reproduce step (Required)

There may be various cases. Here are two examples.

#### Case 1: RC transaction uses IndexReader when the index is dropped and before GC.

```sql
create table t (id int primary key, v int, index iv (v));
insert into t values (1, 10), (2, 20), (3, 30), (4, 40);
```

| session1 | session2 | result |
|--|--|--|
| `set transaction isolation level read committed; begin;` | | |
| `explain select * from t where v = 10;` | | `IndexReader/IndexRangeScan` |
| `select * from t where v = 10;` | | `(1, 10)` |
| | `alter table t drop index iv;` | |
| | `update t set v = 11 where id = 1;` | |
| `explain select * from t where v = 10;` | | `IndexReader/IndexRangeScan` |
| `select * from t where v = 10;` | | `(1, 10)` |
| `select * from t where id = 1;` | | `(1, 11)` |

#### Case 2: RC transactionuses IndexLookup when the index is dropped and before GC.

```sql
create table t (id int primary key, v int, index iv (v), v2 int);
insert into t values (1, 10, 100), (2, 20, 200), (3, 30, 300), (4, 40, 400);
```

| session1 | session2 | result |
|--|--|--|
| `set transaction isolation level read committed; begin;` | | |
| `select * from t use index (iv) where v = 10;` | | `(1, 10, 100)` |
| | `alter table t drop index iv;` | |
| | `update t set v = 11 where id = 1;` | |
| `select * from t use index (iv) where v = 10;` | | `(1, 11, 100)` |
| | `update t set id = 5 where id = 1;` | |
| `select * from t use index (iv) where v = 10;` | | `ERROR 1105 (HY000): inconsistent index iv handle count 1 isn't equal to value count 0` |

### 2. What did you expect to see? (Required)

For case 1, there last `select * from t where v = 10;` should return empty result;
For case 2, session 1 should be able to use index (iv) in the last query

### 3. What did you see instead (Required)

As listed in the table above.

### 4. What is your TiDB version? (Required)

master (3a32bd2df58d3f85504b8f928c70787ced75333a), compiled with fail point (but not used)

### PR DESCRIPTION
**Pull request by you06**
### What problem does this PR solve?

cherry pick #21596 to master as a temporary solution.

Issue Number: close #21498

Problem Summary:

ForUpdateRead will read data using schema version of startTS, this may lead to incorrectness, the cases are listed in #21498

### What is changed and how it works?

What's Changed:

For RC and forUpdateRead in RR, check if index's state is still `StatePublic`.

How it Works:

This check confirms index between snapshotTS and forUpdateTS is not changed, otherwise, the index should not be used.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU, there is a little regression testing with tpc-c

### Release note 

- Fix a bug that schema change will lead to incorrectness when forUpdateRead

### CODE REVIEW

**Review by jackysp — COMMENTED** [review]
LGTM

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -3290,12 +3290,21 @@ func (b *PlanBuilder) buildSelect(ctx context.Context, sel *ast.SelectStmt) (p L
...
+
+	if sel.From != nil {
+		// For sub-queries, the FROM clause may have already been built in outer query when resolving correlated aggregates.
+		// If the ResultSetNode inside FROM clause has nothing to do with correlated aggregates, we can simply get the
+		// existing ResultSetNode from the cache.
+		p, err = b.buildTableRefsWithCache(ctx, sel.From)
+		if err != nil {
+			return nil, err
+		}
+	} else {
```
Comment (lysu) [inline planner/core/logical_plan_builder.go:3306]: why add this `if else`~?

it seems `buildTableRefsWithCache` will return tableDual when `sel.From == nil` 

https://github.com/pingcap/tidb/blob/b45fa8b891b80376b37ab39759730310a182ae5a/planner/core/logical_plan_builder.go#L309

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -3599,7 +3608,7 @@ func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, as
 	if tblName.L == "" {
 		tblName = tn.Name
 	}
-	possiblePaths, err := getPossibleAccessPaths(b.ctx, b.TableHints(), tn.IndexHints, tbl, dbName, tblName)
+	possiblePaths, err := getPossibleAccessPaths(b.ctx, b.TableHints(), tn.IndexHints, tbl, dbName, tblName, b.isForUpdateRead, b.is.SchemaMetaVersion())
```
Comment (lysu) [inline planner/core/logical_plan_builder.go:3611]: `buildTableRefsWithCache`'s cache mechanism should be take care...

for sql: 

```
create table t1(id int);
create table t2(id int);
select (select id from t2 limit 1 for update), id from t1;
```

the value in this line should be:

|  tbl   | b.isForUpdateRead  |
|  ----  | ----  |
| t1  | false |
| t2  | true |

but current got:

|  tbl   | b.isForUpdateRead  |
|  ----  | ----  |
| t1  | false |
| t2  | false |

Comment (you06) [reply planner/core/logical_plan_builder.go:3611]: Good catch, added some cases for `resolveSelect`.

**Comment by lysu** [comment]
LGTM

**Comment by you06** [comment]
I have a question about forUpdateRead.

```sql
insert into t1(id, v, v2) select (select id from t where v = 33), v, v2 from t where v = 34;
select (select id from t where v = 33), v, v2 from t where v = 34 for update;
```
For the above SQLs, `select *** where v = 34` should use for update read, but what about the sub query of `select *** where v = 33`?

File: planner/core/planbuilder.go [hunk 3]
```diff
@@ -892,7 +898,34 @@ func fillContentForTablePath(tablePath *util.AccessPath, tblInfo *model.TableInf
...
 
-func getPossibleAccessPaths(ctx sessionctx.Context, tableHints *tableHintInfo, indexHints []*ast.IndexHint, tbl table.Table, dbName, tblName model.CIStr) ([]*util.AccessPath, error) {
+func isForUpdateReadSelectLock(lock *ast.SelectLockInfo) bool {
+	if lock == nil {
+		return false
+	}
+	return lock.LockType == ast.SelectLockForUpdate || lock.LockType == ast.SelectLockForUpdateNoWait
+}
+
+func getLatestIndexInfo(ctx sessionctx.Context, id int64, startVer int64) (map[int64]*model.IndexInfo, bool, error) {
```
Comment (cfzjywxk) [inline planner/core/planbuilder.go:908]: We could add some comments to explain what does latest mean.

**Comment by lysu** [comment]
> I have a question about forUpdateRead.
> 
> ```sql
> insert into t1(id, v, v2) select (select id from t where v = 33), v, v2 from t where v = 34;
> select (select id from t where v = 33), v, v2 from t where v = 34 for update;
> ```
> 
> For the above SQLs, `select *** where v = 34` should use for update read, but what about the sub query of `select *** where v = 33`?

for mysql, `...where v = 34` should be current-read, and `...where v = 33` should be snapshot-read, so 

```
create table t (id int primary key, v int);
insert into t values(1, 1), (2, 2);

-- t1: begin;
-- t1: select v from t where id in (1, 2) ----> v=1
-- t2: update t set v = v * 10 where id in (1, 2)
-- t1: select v, (select v from t where id = 1) vv from t where id = 2 for update  --> v = 20, vv = 1
```

**Comment by cfzjywxk** [comment]
> I have a question about forUpdateRead.
> 
> ```sql
> insert into t1(id, v, v2) select (select id from t where v = 33), v, v2 from t where v = 34;
> select (select id from t where v = 33), v, v2 from t where v = 34 for update;
> ```
> 
> For the above SQLs, `select *** where v = 34` should use for update read, but what about the sub query of `select *** where v = 33`?

@lysu For TiDB I remember the subquery always use snapshot read, it that right?

**Comment by lysu** [comment]
it seems tidb work well for those test in decorrelated-subquery, but maybe it will break when correlate-subquery(for example: be rewriten into a join) :joy:

lock-read in subquery is corner case...- -
```
MySQL [test]> select v, (select v from t where id = 1 for update) vv from t where id = 2 for update;
+------+------+
| v    | vv   |
+------+------+
|   20 |   10 |
+------+------+
1 row in set (0.003 sec)

MySQL [test]> 
MySQL [test]> 
MySQL [test]> 
MySQL [test]> select v, (select v from t where id = 1 for update) vv from t where id = 2;
+------+------+
| v    | vv   |
+------+------+
|    2 |   10 |
+------+------+
1 row in set (0.002 sec)

MySQL [test]> select v, (select v from t where id = 1) vv from t where id = 2 for update;
+------+------+
| v    | vv   |
+------+------+
|   20 |    1 |
+------+------+
1 row in set (0.003 sec)

MySQL [test]> select v, (select v from t where id = 1) vv from t where id = 2;
+------+------+
| v    | vv   |
+------+------+
|    2 |    1 |
+------+------+
1 row in set (0.003 sec)

MySQL [test]> select tidb_version();
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                              |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: None
Edition: Community
Git Commit Hash: None
Git Branch: None
UTC Build Time: None
GoVersion: go1.15
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.001 sec)
```

**Comment by cfzjywxk** [comment]
> it seems tidb work well for those test in decorrelated-subquery, but maybe it will break when correlate-subquery(for example: be rewriten into a join) 😂
> 
> lock-read in subquery is corner case...- -
> 
> ```
> MySQL [test]> select v, (select v from t where id = 1 for update) vv from t where id = 2 for update;
> +------+------+
> | v    | vv   |
> +------+------+
> |   20 |   10 |
> +------+------+
> 1 row in set (0.003 sec)
> 
> MySQL [test]> 
> MySQL [test]> 
> MySQL [test]> 
> MySQL [test]> select v, (select v from t where id = 1 for update) vv from t where id = 2;
> +------+------+
> | v    | vv   |
> +------+------+
> |    2 |   10 |
> +------+------+
> 1 row in set (0.002 sec)
> 
> MySQL [test]> select v, (select v from t where id = 1) vv from t where id = 2 for update;
> +------+------+
> | v    | vv   |
> +------+------+
> |   20 |    1 |
> +------+------+
> 1 row in set (0.003 sec)
> 
> MySQL [test]> select v, (select v from t where id = 1) vv from t where id = 2;
> +------+------+
> | v    | vv   |
> +------+------+
> |    2 |    1 |
> +------+------+
> 1 row in set (0.003 sec)
> 
> MySQL [test]> select tidb_version();
> +---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | tidb_version()                                                                                                                                                                                                                              |
> +---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | Release Version: None
> Edition: Community
> Git Commit Hash: None
> Git Branch: None
> UTC Build Time: None
> GoVersion: go1.15
> Race Enabled: false
> TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
> Check Table Before Drop: false |
> +---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> 1 row in set (0.001 sec)
> ```

That is a problem..  Actually if two statements are equal from the planner prospective but returns different results..
I'm confused.. How should we set our expected behaviours?

**Comment by cfzjywxk** [comment]
Seems here it's coupled with planner much more than we think for the `for update` statement processing.

**Comment by you06** [comment]
From my test in mysql 5.7.32

```
insert into t1(id, v, v2) select (select id from t where v = 33), v, v2 from t where v = 34;
                                           └ current read                   └ current read
select (select id from t where v = 33), v, v2 from t where v = 34 for update;
                      └ snapshot read             └ current read
```

945f4cf forbid the propagate of current read for correlate-aggregate query, in another word, correlate-subquery should use snapshot read unless it has forUpdate decorated, we got the following result then.

```
insert into t1(id, v, v2) select (select id from t where v = 33), v, v2 from t where v = 34;
                                           └ snapshot read                   └ current read
select (select id from t where v = 33), v, v2 from t where v = 34 for update;
                      └ snapshot read             └ current read
```

@lysu, @cfzjywxk PTAL.

**Comment by cfzjywxk** [comment]
> [945f4cf](https://github.com/pingcap/tidb/commit/945f4cf86d47c9ca98d5e3aa78a8a38e4feebe61) forbid the propagate of current read for correlate-aggregate query, in another word, correlate-subquery should use snapshot read unless it has forUpdate decorated, we got the following result then.
> 
> ```
> insert into t1(id, v, v2) select (select id from t where v = 33), v, v2 from t where v = 34;
>                                            └ snapshot read                   └ current read
> select (select id from t where v = 33), v, v2 from t where v = 34 for update;
>                       └ snapshot read             └ current read
> ```
> 
> @lysu, @cfzjywxk PTAL.

@you06 
This place should be rewriting the subquery expression but not `correlatedAggregateResolver`?

**Comment by you06** [comment]
> > [945f4cf](https://github.com/pingcap/tidb/commit/945f4cf86d47c9ca98d5e3aa78a8a38e4feebe61) forbid the propagate of current read for correlate-aggregate query, in another word, correlate-subquery should use snapshot read unless it has forUpdate decorated, we got the following result then.
> > ```
> > insert into t1(id, v, v2) select (select id from t where v = 33), v, v2 from t where v = 34;
> >                                            └ snapshot read                   └ current read
> > select (select id from t where v = 33), v, v2 from t where v = 34 for update;
> >                       └ snapshot read             └ current read
> > ```
> > 
> > 
> > @lysu, @cfzjywxk PTAL.
> 
> @you06
> This place should be rewriting the subquery expression but not `correlatedAggregateResolver`?

Yes, and I added the case, the capture mechanism should be patches to release-4.0 also.

**Comment by cfzjywxk** [comment]
@lysu PTAL

File: planner/core/expression_rewriter.go [hunk 4]
```diff
@@ -107,7 +107,11 @@ func (b *PlanBuilder) rewriteInsertOnDuplicateUpdate(ctx context.Context, exprNo
 // asScalar means whether this expression must be treated as a scalar expression.
 // And this function returns a result expression, a new plan that may have apply or semi-join.
 func (b *PlanBuilder) rewrite(ctx context.Context, exprNode ast.ExprNode, p LogicalPlan, aggMapper map[*ast.AggregateFuncExpr]int, asScalar bool) (expression.Expression, LogicalPlan, error) {
+	// capture the forUpdateRead mark
+	isForUpdateRead := b.isForUpdateRead
+	b.isForUpdateRead = false
 	expr, resultPlan, err := b.rewriteWithPreprocess(ctx, exprNode, p, aggMapper, nil, asScalar, nil)
+	b.isForUpdateRead = isForUpdateRead
```
Comment (lysu) [inline planner/core/expression_rewriter.go:114]: it seems without this change, TestIssue21498 still can work

and it seems subQuery in Projection will call `rewriteWithPreprocess` directly

https://github.com/pingcap/tidb/blob/f6cac2f831e54643d4367f6ab7a206f9ec7e1dcf/planner/core/logical_plan_builder.go#L1222

File: planner/core/logical_plan_builder.go [hunk 5]
```diff
@@ -2253,6 +2253,10 @@ func (r *correlatedAggregateResolver) resolveSelect(sel *ast.SelectStmt) (err er
 	if err != nil {
 		return err
 	}
+	// do not use cache when for update read
+	if isForUpdateReadSelectLock(sel.LockInfo) {
+		useCache = false
```
Comment (lysu) [inline planner/core/logical_plan_builder.go:2258]: after debug, it seems `useCache = false` only not-use-cache, but it still refill cache..

so in Access Path Selection will happen in `correlatedAggregateResolver` and directly use cache in later buildProjection

**Comment by lysu** [comment]
@you06 @cfzjywxk how about fix subQuery question in other PRs, it seems too many question in subquery + lock like https://github.com/pingcap/tidb/issues/22345, but it's not critial

File: planner/core/logical_plan_builder.go (hunk 5 above)
Comment (you06) [reply planner/core/logical_plan_builder.go:2258]: Seems `useCache = false` only cache sub-sub-query when there is a select stmt inside from clause.

File: planner/core/expression_rewriter.go (hunk 4 above)
Comment (you06) [reply planner/core/expression_rewriter.go]: Yeah, removed the capture now, at least not in this PR.

**Comment by lysu** [comment]
LGTM

File: planner/core/logical_plan_builder.go (hunk 5 above)
Comment (you06) [reply planner/core/logical_plan_builder.go:2262]: I added some cases for sub queries, the cache problem is also related with rewrite, it may be solved together with it.

**Comment by cfzjywxk** [comment]
LGTM

**Review by ti-srebot — APPROVED** [review]
